tools/flash_test/flash_test.c

```diff
@@ -0,0 +1,201 @@
+#include <stdio.h>
+#include <stdint.h>
+
+#include "../hidapi.h"
+#include "../hidapi.c"
+
+#define VID 0x303a
+#define PID 0x4004
+
+hid_device * hd;
+const int chunksize = 128;
+int tries = 0;
+int alignlen = 4;
+
+uint8_t rdata[256];
+int sector_size = 4096;
+
+int StringToNumberHexOrInt( const char * str )
+{
+	if( !str ) return 0;
+	if( strlen( str ) > 1 )
+	{
+		if( str[0] == '0' && str[1] == 'x' )
+		{
+			return strtol( str+2, 0, 16 );
+		}
+	}
+	return atoi( str );
+}
+
+int main( int argc, char ** argv )
+{
+	if( argc != 4 && argc != 5 )
+	{
+		goto help;
+	}
+	
+	uint32_t flash_start = StringToNumberHexOrInt( argv[3] );
+		
+	hid_init();
+	hd = hid_open( VID, PID, 0);
+	if( !hd ) { fprintf( stderr, "Could not open USB\n" ); return -94; }
+
+	if( strcmp( argv[1], "write" ) == 0 )
+	{
+		if( argc != 4 ) goto help;
+		struct stat sbuf = { 0 };
+		int r = stat( argv[2], &sbuf);
+		int bloblen = sbuf.st_size;
+		if( r < 0 )
+		{
+			fprintf( stderr, "Could not open file\n" );
+			return -1;
+		}
+		if( bloblen == 0 )
+		{
+			fprintf( stderr, "No file to write\n" );
+			return -1;
+		}
+		
+		printf( "Writing file %s to 0x%x (length: %d)\n", argv[2], flash_start, bloblen );
+		int blobkclen = ( bloblen + alignlen - 1 ) / alignlen * alignlen;
+		uint8_t * blob;
+		{
+			FILE * binary_blob = fopen( argv[2], "r" );
+			if( !binary_blob )
+			{
+				fprintf( stderr, "ERROR: Could not open %s.\n", argv[2] );
+				return -96;
```

Comment:
Where are -96 and -97 coming from?

Reply:
Ok. I always prefer something marginally less arbitrary, like an enum, but it is what it is.